src/System.Net.Security/tests/FunctionalTests/SslStreamSNITest.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Net.Test.Common;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Security.Tests
+{
+    using Configuration = System.Net.Test.Common.Configuration;
+
+    public class SslStreamSNITest
+    {
+        private static IEnumerable<object[]> HostNameData()
+        {
+            yield return new object[] { "a" };
+            yield return new object[] { "test" };
+            yield return new object[] { new string('a', 100) };
+        }
+
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            WithVirtualConnection((server, client) =>
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                });
+
+                SslServerAuthenticationOptions options = DefaultServerOptions();
+
+                int timesCallbackCalled = 0;
+                options.ServerCertificateSelectionCallback = (sender, actualHostName) =>
+                {
+                    timesCallbackCalled++;
+                    Assert.Equal(hostName, actualHostName);
+                    return serverCert;
+                };
+
+                var cts = new CancellationTokenSource();
+                server.AuthenticateAsServerAsync(options, cts.Token).Wait();
+
+                Assert.Equal(1, timesCallbackCalled);
+                clientJob.Wait();
+            },
+            (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+                {
+                    Assert.Equal(serverCert, certificate);
+                    return true;
+                });
+        }
+
+        [Fact]
+        public void SslStream_UnknownHostName_Fails()
```

Comment:
Might be better to rename this test to indicate that it is testing that no SNI data was sent from client to server and thus the callback returns null.

So, perhaps this is a better name:  "SslStream_NoSniFromClient_CallbackReturnsNull"

Reply:
@davidsh, thanks, I thought the last part is always Fails or Ok 😄

Reply:
The last part is either "Success".  Some folks use "Succeeds".  But "Success" is better. IMHO.

For failure cases, it always better to use terms like "Throws" or "ThrowsSocketException", or in this case, "CallbackReturnsNull".  It is more descriptive and helps us understand the test case better.